editor/src/core/shared/element-path-tree.ts

```diff
@@ -25,57 +31,160 @@ export function elementPathTree(
+  const otherChildrenFromPaths = paths.filter(
+    (path) =>
+      EP.isChildOf(path, rootPath) &&
+      !childrenFromElement.some((other) => EP.pathsEqual(other, path)),
+  )
+
+  // If there are children outside the element metadata, need to merge the two and sort them.
+  // Otherwise, return the children from the meta.
+  return otherChildrenFromPaths.length > 0
+    ? childrenFromElement
+        .concat(otherChildrenFromPaths)
+        .sort(
+          (a, b) =>
+            paths.findIndex((p) => EP.pathsEqual(p, a)) -
+            paths.findIndex((p) => EP.pathsEqual(p, b)),
+        )
+    : childrenFromElement
+}
 
-export function buildTree(elementPaths: Array<ElementPath>): ElementPathTrees {
-  let result: ElementPathTrees = {}
-  let currentPathParts: Array<ElementPathPart> = []
-  let priorPathString: string | null = null
-  // Loop over each and every path that needs to be added.
-  for (let elementPathIndex = 0; elementPathIndex < elementPaths.length; elementPathIndex++) {
-    const elementPath = elementPaths[elementPathIndex]
-    currentPathParts = []
-    priorPathString = null
-    // Loop over the array segments of the path that is being added.
-    for (let pathPartIndex = 0; pathPartIndex < elementPath.parts.length; pathPartIndex++) {
-      const pathPart = elementPath.parts[pathPartIndex]
-      currentPathParts[pathPartIndex] = []
-      let currentPathPart = currentPathParts[pathPartIndex]
-      // Loop over the parts of the of the path segment.
-      for (
-        let pathPartElementIndex = 0;
-        pathPartElementIndex < pathPart.length;
-        pathPartElementIndex++
-      ) {
-        const pathPartElement = pathPart[pathPartElementIndex]
-        currentPathPart.push(pathPartElement)
-        const subElementPath = EP.elementPath(currentPathParts)
-
-        // See if there's an already existing part of the tree with this path.
-        const subElementPathString = EP.toString(subElementPath)
-        const foundTree = result[subElementPathString]
-        let treeToTarget = foundTree ?? emptyElementPathTree(subElementPath, subElementPathString)
-
-        // Add this if it doesn't already exist.
-        if (foundTree == null) {
-          result[subElementPathString] = treeToTarget
-          // Handle adding this entry to its parent as a child.
-          if (priorPathString != null) {
-            const childrenOfPriorPath = result[priorPathString].children
-            childrenOfPriorPath.push(treeToTarget)
+function getMissingParentPaths(
+  paths: ElementPath[],
+  metadata: ElementInstanceMetadataMap,
+): ElementPath[] {
+  const missingParentPaths = new Set<ElementPath>()
+  for (const path of paths) {
+    let parent = EP.parentPath(path)
+    while (parent.parts.length > 0) {
+      if (metadata[EP.toString(parent)] == null) {
+        missingParentPaths.add(parent)
+      }
+      parent = EP.parentPath(parent)
+    }
+  }
+  return [...missingParentPaths.values()]
```

Comment:
I _think_ you can do
```suggestion
  return Array.from(missingParentPaths)
```
to avoid an intermediary value.

Reply:
https://github.com/concrete-utopia/utopia/pull/3811/commits/5337e065eee25e8558f5c59be1df873d9a751333